pyonfx/timestamps.py

```diff
@@ -0,0 +1,552 @@
+
+        Parameters:
+            file_content (TextIOWrapper): The timestamps content
+            version (int): The version of the timestamps (only 2 or 4 is allowed)
+            rounding_method (RoundingMethod): A rounding method
+
+        Returns:
+            A tuple containing these 3 informations:
+                1. A list of each timestamps rounded to milliseconds
+                2. The last timestamps not rounded
+                3. The fpms
+        """
+
+        if version != 2 and version != 4:
+            raise ValueError("You can only specify version 2 or 4.")
+
+        timestamps: List[int] = []
+        previous_timestamp: int = 0
+        lowest_timestamp: Fraction = None
+        highest_timestamp: Fraction = None
+
+        for line in file_content:
+            line = line.strip(" \t")
+
+            if not line or line.startswith("#"):
+                continue
+
+            try:
+                timestamp = Fraction(line)
+            except ValueError:
+                raise ValueError(
+                    f'The timestamps file contain a invalid line. Here is it: "{line}"'
+                )
+
+            if highest_timestamp is None or highest_timestamp < timestamp:
+                highest_timestamp = timestamp
+            if lowest_timestamp is None or lowest_timestamp > timestamp:
+                lowest_timestamp = timestamp
+
+            rounded_timestamp = rounding_method(timestamp)
+
+            if version == 2 and rounded_timestamp < previous_timestamp:
+                raise ValueError(
+                    f"The timestamps file contain timestamps NOT in ascending order."
+                )
+
+            previous_timestamp = rounded_timestamp
+            timestamps.append(rounded_timestamp)
+
+        if not len(timestamps):
+            raise ValueError(f"The timestamps file is empty.")
+
+        if version == 4:
+            timestamps.sort()
+
+        duration = highest_timestamp - lowest_timestamp
+        if duration:
+            fpms = Fraction(len(timestamps) - 1) / (
+                highest_timestamp - lowest_timestamp
+            )
+        else:
+            fpms = 0
+        return timestamps, highest_timestamp, fpms
+
+
+class Timestamps:
+    """Timestamps object contains informations about the timestamps of an video.
+
+    Video player have 2 methods to deal with timestamps. Some floor them and other round them.
+    This can lead to difference when displaying the subtitle.
+        Ex:
+            Player - Method - proof
+            mpv    - round  - https://github.com/mpv-player/mpv/blob/7480efa62c0a2a1779b4fdaa804a6512aa488400/sub/sd_ass.c#L499
+            FFmpeg - floor  - https://github.com/FFmpeg/FFmpeg/blob/fd1712b6fb8b7acc04ccaa7c02b9a5c9f233cfb3/libavfilter/vf_subtitles.c#L195
+            VLC    - floor  - https://github.com/videolan/vlc/blob/f7bb59d9f51cc10b25ff86d34a3eff744e60c46e/include/vlc_tick.h#L118-L130
```

Comment:
First of all, I should have used the gitlab repos, not github since I think their principal git repos is gitlab.
They call ass_render_frame [here](https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/modules/codec/libass.c#L453-454) which call [MS_FROM_VLC_TICK](https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/include/vlc_tick.h#L120-132)

I don't fully understand what a tick is, but here is my understanding: A [tick](https://code.videolan.org/videolan/vlc/-/blob/df6394ea8003e035a281b6818e6432c7d492ed2f/include/vlc_tick.h#L36-46) is a unit of time greater than milliseconds (seems to be microseconds). As you may know, double have a floating-point precision problem, so I assume VLC uses a higher unit to achieve better precision. Now, as for why it floors, it's because both variables (``vtk`` and ``CLOCK_FREQ``) are integers, which means the result will be truncated.

Reply:
Gotcha, seems legit :)

Reply:
I updated VLC link in the documentation
Hope your ok with that